task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,230 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.constant.ServiceError;
+import com.epam.esm.dto.GiftCertificateDto;
+import com.epam.esm.converter.Converter;
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.repository.GiftCertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.repository.exception.RepositoryException;
+import com.epam.esm.service.GiftCertificateService;
+import com.epam.esm.service.exception.certificate.GiftCertificateAlreadyExistException;
+import com.epam.esm.service.exception.certificate.GiftCertificateNotFoundException;
+import com.epam.esm.service.exception.ServiceException;
+import com.epam.esm.service.exception.certificate.UnableDeleteGiftCertificateException;
+import com.epam.esm.service.exception.certificate.UnableUpdateGiftCertificate;
+import com.epam.esm.util.ParamsUtil;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+    private static final Logger logger = LogManager.getLogger(GiftCertificateServiceImpl.class);
+
+    @Autowired
+    private GiftCertificateRepository giftCertificateRepository;
+    @Autowired
+    private TagRepository tagRepository;
+    @Autowired
+    private Converter<GiftCertificate, GiftCertificateDto> giftCertificateConverter;
+
+    @Override
+    public List<GiftCertificateDto> findAllWithTags(Map<String, String[]> params) throws ServiceException {
+        try {
```

Comment:
@igor-zhuravlev It would be nice to use this construct elsewhere which defines the calling method